### PR TITLE
Prowler finding alert

### PR DESCRIPTION
**Pull request by andrew-kline**
- Lambda for sending slack alerts
- DDB stream from findings table to trigger Lambda
- Lambda role to read from DDB and from config file in S3

### CODE REVIEW

File: cloudformation/Prowler-Template.yaml [hunk 1]
```diff
@@ -131,6 +131,7 @@ Parameters:
 Conditions:
   cSlackSecret: !Not [!Equals [ !Ref pSlackSecret, "NONE" ]]
   cDeployIngestLambda: !Not [!Equals [ !Ref pDomainEndpoint, "NONE" ]]
+  cDeployAlertLambda: !Not [!Equals [ !Ref pSlackSecret, "NONE" ]]
```
Comment (jchrisfarris) [inline cloudformation/Prowler-Template.yaml]: See my comment in the other repo - The pSlackSecret will do two things that should be controlled independently.

File: cloudformation/Prowler-Template.yaml [hunk 2]
```diff
@@ -679,6 +680,85 @@ Resources:
...
+      Handler: alert_prowler_findings.handler
+      Role: !GetAtt FindingsAlertRole.Arn
+      CodeUri: ../lambda
+      Timeout: 60
+      Environment:
+        Variables:
+          SLACK_SECRET: !Sub "arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:${pSlackSecret}"
+          CONFIG_BUCKET: !Ref ProwlerBucket
+      Layers:
+        - arn:aws:lambda:us-east-1:177933569100:layer:AWS-Parameters-and-Secrets-Lambda-Extension:11
```
Comment (jchrisfarris) [inline cloudformation/Prowler-Template.yaml]: What is this layer?

File: lambda/alert_prowler_findings.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+if not config_data:
+    logger.warning("Config data/file not found. No alerts will be sent")
+    config_data = {}
+
+logger.info("alert_prowler_findings initiated")
+
+# Lambda execution starts here
+def handler(event, context):
+    logger.debug("Received event: " + json.dumps(event, sort_keys=True))
+    slack_secret = get_cache_secret(SLACK_SECRET_ARN)
```
Comment (jchrisfarris) [inline lambda/alert_prowler_findings.py]: We should move that outside the handler with the S3 download of the Config

File: lambda/common.py [hunk 4]
```diff
@@ -93,6 +99,37 @@ def get_org_account_details():
         else:
             raise
 
+def get_cache_secret(secret_arn):
```
Comment (jchrisfarris) [inline lambda/common.py]: What's happening here?

Comment (jchrisfarris) [reply lambda/common.py]: I guess my confusion is - this is not the normal way to get a secret and shared in the AWS console.

**Review by jchrisfarris — CHANGES_REQUESTED** [review]

File: cloudformation/Prowler-Template.yaml (hunk 1 above)
Comment (andrew-kline) [reply cloudformation/Prowler-Template.yaml]: updated to use unique parameter

File: cloudformation/Prowler-Template.yaml (hunk 2 above)
Comment (andrew-kline) [reply cloudformation/Prowler-Template.yaml]: basically in-memory caching for secrets. reduces calls to SM - faster and cheaper. It is an [AWS-provided layer](https://docs.aws.amazon.com/systems-manager/latest/userguide/ps-integration-lambda-extensions.html#ps-integration-lambda-extensions-add). [Blog post](https://aws.amazon.com/blogs/compute/using-the-aws-parameter-and-secrets-lambda-extension-to-cache-parameters-and-secrets/) and [doc](https://docs.aws.amazon.com/secretsmanager/latest/userguide/retrieving-secrets_lambda.html) describing how it works.

File: lambda/common.py (hunk 4 above)
Comment (andrew-kline) [reply lambda/common.py]: see the links above IRT the layer - this should cache the cred in-memory to reduce calls to SM. I like this design for cost and speed reduction - but I can go the route of calling SM directly, just let me know

File: cloudformation/Prowler-Template.yaml (hunk 2 above)
Comment (andrew-kline) [reply cloudformation/Prowler-Template.yaml]: couple other use examples https://github.com/aws-samples/serverless-patterns/blob/8d77bb0725e257c32bb84b8e1ead9b0bd2ff600b/lambda-secrets-manager-extension-python/hello_world/app.py#L10 and https://github.com/jfrog/artifactory-sagemaker/blob/main/infrastructure/lambda_function.py

Comment (jchrisfarris) [reply cloudformation/Prowler-Template.yaml]: Makes total sense, and this is a nifty solution. 

But, in this  case we're pulling just the Slack Secret, which will rarely change. This probably doesn't need the performance boost of the caching. 

Rather I'd be concerned around the usage of an unknown layer in my client's environment, and the same questions I asked you being asked of me.

File: lambda/common.py (hunk 4 above)
Comment (jchrisfarris) [reply lambda/common.py]: I think I'd prefer the standard method to fetch the slack secret at execution environment creation.

**Review by jchrisfarris — CHANGES_REQUESTED** [review]
Lets do the normal secret fetching process instead.

**Comment by andrew-kline** [comment]
> Lets do the normal secret fetching process instead.

added

**Review by jchrisfarris — APPROVED** [review]

**Pull request merged**